mr/src/main/java/org/apache/iceberg/mr/Catalogs.java

```diff
@@ -180,47 +191,82 @@ public static boolean dropTable(Configuration conf, Properties props) {
-        case HADOOP:
+    Map<String, String> properties = getCatalogProperties(conf, catalogName);
+    Catalog catalog;
+    switch (catalogType.toLowerCase()) {
+      case HADOOP:
+        if (properties.containsKey(CatalogProperties.WAREHOUSE_LOCATION)) {
+          catalog = CatalogUtil.loadCatalog(HadoopCatalog.class.getName(), catalogName, properties, conf);
+        } else {
           String warehouseLocation = conf.get(InputFormatConfig.HADOOP_CATALOG_WAREHOUSE_LOCATION);
-
           catalog = (warehouseLocation != null) ? new HadoopCatalog(conf, warehouseLocation) : new HadoopCatalog(conf);
-          LOG.info("Loaded Hadoop catalog {}", catalog);
+        }
+        LOG.info("Loaded Hadoop catalog {}", catalog);
+        return Optional.of(catalog);
+      case HIVE:
+        catalog = HiveCatalogs.loadCatalog(catalogName, properties, conf);
+        LOG.info("Loaded Hive Metastore catalog {}", catalog);
+        return Optional.of(catalog);
+      case CUSTOM:
+        if (properties.containsKey(CatalogProperties.CATALOG_IMPL)) {
+          String catalogLoaderClass = properties.get(CatalogProperties.CATALOG_IMPL);
+          catalog = CatalogUtil.loadCatalog(catalogLoaderClass, catalogName, properties, conf);
+          LOG.info("Loaded catalog {} using {}", catalog, catalogLoaderClass);
           return Optional.of(catalog);
-        case HIVE:
-          catalog = HiveCatalogs.loadCatalog(conf);
-          LOG.info("Loaded Hive Metastore catalog {}", catalog);
+        } else if (conf.get(InputFormatConfig.CATALOG_LOADER_CLASS) != null) {
+          String catalogLoaderClass = conf.get(InputFormatConfig.CATALOG_LOADER_CLASS);
+          CatalogLoader loader = (CatalogLoader) DynConstructors.builder(CatalogLoader.class)
+                  .impl(catalogLoaderClass)
+                  .build()
+                  .newInstance();
+          catalog = loader.load(conf);
+          LOG.info("Loaded catalog {} using {}", catalog, catalogLoaderClass);
           return Optional.of(catalog);
-        default:
-          throw new NoSuchNamespaceException("Catalog %s is not supported.", catalogName);
-      }
+        } else {
+          return Optional.empty();
+        }
+      default:
+        throw new NoSuchNamespaceException("Catalog %s is not supported.", catalogType);
     }
+  }
 
-    LOG.info("Catalog is not configured");
-    return Optional.empty();
+  private static Map<String, String> getCatalogProperties(Configuration conf, String catalogName) {
+    String keyPrefix = InputFormatConfig.CATALOG_CONFIG_PREFIX + catalogName;
+    return Streams.stream(conf.iterator()).filter(e -> e.getKey().startsWith(keyPrefix))
+            .collect(Collectors.toMap(e -> e.getKey().substring(keyPrefix.length() + 1), Map.Entry::getValue));
+  }
+
+  private static String getCatalogType(Configuration conf, String catalogName) {
+    String name = catalogName == null ? DEFAULT_CATALOG : catalogName;
```

Comment:
Why fill in the default catalog name if this is going to fall back to the global configuration? It seems like we need one fallback mechanism, not two. Are you trying to replace the current fallback?

Reply:
Okay, after going back to the description, it looks like that is the intent.